src/ocean/meta/values/VisitValue.d

```diff
@@ -0,0 +1,183 @@
+{
+    VisitImpl!(Visitor) impl;
+    impl.visitor = &visitor;
+    impl.visitAll(&value);
+}
+
+/*******************************************************************************
+
+    Implementation of visiting logic. Some types are currently not supported
+    because it is not clear what would recursing mean for them.
+
+    Params:
+        Visitor = user-provided struct implementing templated `visit` method
+            to be called for individual values
+
+*******************************************************************************/
+
+private struct VisitImpl ( Visitor )
+{
+    /// Refers to the persistent instance of visitor struct in case
+    /// implementation / may need internal state
+    Visitor* visitor;
+
+    /***************************************************************************
+
+        Main recursive type/value visiting implementation
+
+        Params:
+            value = pointer to currently processed value. Pointer is used
+                instead of a reference because D1 does not allow defining
+                a reference to a static array
+
+    ***************************************************************************/
+
+    private void visitAll ( T ) ( T* value )
+    {
+        auto deeper = this.visitor.visit(value);
+
+        if (!deeper)
+            return;
+
+        static if (isPrimitiveType!(T))
+        {
+            // do nothing, already processed
+        }
+        else static if (isTypedef!(T))
+        {
+            auto reinterp = cast(TypedefBaseType!(T)*) value;
+            this.visitAll(reinterp);
+        }
+        else static if (isAggregateType!(T))
+        {
+            static if (is(T == class))
+            {
+                bool ignore = *value is null;
+
+                // Recurse into super class fields
+                static if (is(T Bases == super))
+                {
+                    foreach (Base; Bases)
+                    {
+                        static if (Base.init.tupleof.length)
+                            this.visitAll(cast(Base*) value);
+                    }
+                }
+            }
+            else
+            {
+                bool ignore = false;
+            }
+
+            if (!ignore)
+            {
+                foreach (ref field; (*value).tupleof)
+                {
+                    this.visitAll(&field);
+                }
+            }
+        }
+        else static if (isArrayType!(T))
+        {
+            static if (
+                   isArrayType!(T) == ArrayKind.Static
+                || isArrayType!(T) == ArrayKind.Dynamic)
+            {
+                foreach (ref elem; *value)
+                    this.visitAll(&elem);
+            }
+            else
+            {
+                static assert (false, "AA currently not supported");
+            }
+        }
+        else static if (isFunctionType!(T))
+        {
+            // ignored for now
```

Comment:
"for now" -- what would be the alternative to ignoring function types (other than rejecting them)?

Reply:
This is exactly why "for now" - I can't imagine any iseful processing semantics, but maybe someone will request one :)